File: skills/audit-knowledge-vault/scripts/knowledge_loom/focus.py
from __future__ import annotations

import re
from pathlib import Path

from .models import Finding
from .pathing import resolve_vault_path

HEADING_RE = re.compile(r"^(#{2,3})\s+(.+?)\s*$")


def _active_items(text: str, section_name: str) -> list[str]:
    in_section = False
    items: list[str] = []
    for line in text.splitlines():
        match = HEADING_RE.match(line)
        if not match:
            continue
        level, title = match.groups()
        if level == "##":
            in_section = title.strip().casefold() == section_name.strip().casefold()
            continue
        if in_section and level == "###":
            items.append(title.strip())
    return items


def check_focus_view(root: Path, name: str, view: dict) -> list[Finding]:
    findings: list[Finding] = []
    relative = view.get("path")
    if not isinstance(relative, str):
        return findings
    path = resolve_vault_path(root, relative)
    if path is None:
        return [Finding("error", "focus.boundary", f"focus view `{name}` resolves outside the vault", relative)]
    if not path.is_file():
        return [Finding("error", "focus.missing", f"focus view `{name}` file is missing", relative)]

    text = path.read_text(encoding="utf-8")
    section = view.get("active_section", "Top of mind")
    if not isinstance(section, str):
        return findings
    items = _active_items(text, section)
    max_top = view.get("max_top", 3)
    max_active = view.get("max_active", max_top)
    if not isinstance(max_top, int) or not isinstance(max_active, int):
        return findings

    if len(items) > max_top:
        findings.append(
            Finding("error", "focus.max-top", f"focus view `{name}` has {len(items)} items; maximum is {max_top}", relative)
        )

    active = [item for item in items if not re.search(r"(?:—|-|\[)\s*next\b", item, flags=re.IGNORECASE)]
    if len(active) > max_active:
        findings.append(
            Finding(
                "error",
                "focus.max-active",
                f"focus view `{name}` has {len(active)} active items; maximum is {max_active}",
                relative,
            )
        )

    if view.get("require_start_here", False):
        starts = [item for item in items if "start here" in item.casefold()]
        if len(starts) != 1:
            findings.append(
                Finding(
                    "error",
                    "focus.start-here",
                    f"focus view `{name}` requires exactly one Start here item; found {len(starts)}",
                    relative,
                )
            )
    return findings

File: src/knowledge-loom/focus.ts
import fs from "node:fs";

import { finding } from "./contract.js";
import { resolveVaultPath } from "./pathing.js";
import type { Finding, UnknownRecord } from "./types.js";

const HEADING_RE = /^(#{2,3})\s+(.+?)\s*$/;

function activeItems(text: string, sectionName: string): string[] {
  let inSection = false;
  const items = [];
  for (const line of text.split(/\r?\n/)) {
    const match = line.match(HEADING_RE);
    if (!match) continue;
    const [, level, title] = match;
    if (!level || !title) continue;
    if (level === "##") {
      inSection = title.trim().toLocaleLowerCase() === sectionName.trim().toLocaleLowerCase();
    } else if (inSection) {
      items.push(title.trim());
    }
  }
  return items;
}

export function checkFocusView(root: string, name: string, view: UnknownRecord): Finding[] {
  const relative = view.path;
  if (typeof relative !== "string") return [];
  const focusPath = resolveVaultPath(root, relative);
  if (focusPath === null) return [finding("error", "focus.boundary", `focus view \`${name}\` resolves outside the vault`, relative)];
  if (!fs.statSync(focusPath, { throwIfNoEntry: false })?.isFile()) {
    return [finding("error", "focus.missing", `focus view \`${name}\` file is missing`, relative)];
  }

  const section = view.active_section ?? "Top of mind";
  if (typeof section !== "string") return [];
  const items = activeItems(fs.readFileSync(focusPath, "utf8"), section);
  const maxTop = view.max_top ?? 3;
  const maxActive = view.max_active ?? maxTop;
  if (typeof maxTop !== "number" || typeof maxActive !== "number" || !Number.isInteger(maxTop) || !Number.isInteger(maxActive)) return [];

  const findings: Finding[] = [];
  if (items.length > maxTop) {
    findings.push(finding("error", "focus.max-top", `focus view \`${name}\` has ${items.length} items; maximum is ${maxTop}`, relative));
  }
  const active = items.filter((item) => !/(?:—|-|\[)\s*next\b/i.test(item));
  if (active.length > maxActive) {
    findings.push(finding("error", "focus.max-active", `focus view \`${name}\` has ${active.length} active items; maximum is ${maxActive}`, relative));
  }
  if (view.require_start_here === true) {
    const starts = items.filter((item) => item.toLocaleLowerCase().includes("start here"));
    if (starts.length !== 1) {
      findings.push(finding("error", "focus.start-here", `focus view \`${name}\` requires exactly one Start here item; found ${starts.length}`, relative));
    }
  }
  return findings;
}
